### PR TITLE
Create service per pod in a statefulset deployment

### PR DESCRIPTION
**Pull request by avinash-platformatory**
In order to expose warpstream agents outside the cluster using host-based routing with TLS, each pod in a statefulset should have a service that the ingress object can point to for every host. This allows to expose each agent in a StatefulSet deployment with a service, allowing agent-aware load balancing instead of load balancing at the load balancer.

With a service created for every pod in a StatefulSet, we can have an ingress that routes host based traffic to specific pods.
Examples of doing it in various ingress controllers -
1. Kong
```yaml
apiVersion: configuration.konghq.com/v1beta1
kind: TCPIngress
metadata:
  name: warpstream-agent-ingress
  namespace: ws
  annotations:
    kubernetes.io/ingress.class: kong
    konghq.com/protocols: tls_passthrough
    konghq.com/snis: "ws-0.ws-headless.ws.svc.example.com, ws-1.ws-headless.ws.svc.example.com, ws-2.ws-headless.ws.svc.example.com"
spec:
  rules:
    - host: ws-0.ws-headless.ws.svc.example.com
      port: 9443
      backend:
        serviceName: ws-0
        servicePort: 9092
    - host: ws-1.ws-headless.ws.svc.example.com
      port: 9443
      backend:
        serviceName: ws-1
        servicePort: 9092
    - host: ws-2.ws-headless.ws.svc.example.com
      port: 9443
      backend:
        serviceName: ws-2
        servicePort: 9092
```
2. Istio Ingress
```yaml
apiVersion: networking.istio.io/v1
kind: VirtualService
metadata:
  name: warpstream-ingress
spec:
  hosts:
  - "ws-0.ws-headless.ws.svc.example.com"
  gateways:
  - warpstream-tls-gateway
  tcp:
  - match:
    - port: 9443
    route:
    - destination:
        host: ws-0
        port:
          number: 9092
  - "ws-1.ws-headless.ws.svc.example.com"
  gateways:
  - warpstream-tls-gateway
  tcp:
  - match:
    - port: 9443
    route:
    - destination:
        host: ws-1
        port:
          number: 9092
  - "ws-2.ws-headless.ws.svc.example.com"
  gateways:
  - warpstream-tls-gateway
  tcp:
  - match:
    - port: 9443
    route:
    - destination:
        host: ws-2
        port:
          number: 9092
```

### CODE REVIEW

File: charts/warpstream-agent/templates/service.yaml
Comment (rmb938) [inline charts/warpstream-agent/templates/service.yaml:37]: can you also add a variable to enable/disable this? Not ever deployment needs per pod services. Let's keep it disabled by default to not change the default behavior.

**Review by rmb938 — CHANGES_REQUESTED** [review]
Hi @avinash-platformatory thanks for the contribution!

Just one comment to add a flag to disable the per pod services. 

I also don't think this will actually work how you are expecting in your case. 

Warpstream agents and Kafka clients won't actually use these services for anything unless the pods are advertising hostnames that point to these services. Each pod would have to have a unique `WARPSTREAM_DISCOVERY_KAFKA_HOSTNAME_OVERRIDE` which isn't possible in the chart.

File: charts/warpstream-agent/templates/service.yaml
Comment (avinash-platformatory) [reply charts/warpstream-agent/templates/service.yaml:37]: Done. Added a flag

**Comment by avinash-platformatory** [comment]
@rmb938 I have tested this against my cluster with the following configuration, and it works since the `WARPSTREAM_DISCOVERY_KAFKA_HOSTNAME_OVERRIDE` is set for a StatefulSet with the pod's name as the prefix - https://github.com/warpstreamlabs/charts/blob/e916ce3c1c0c5977219cdafd4af2221c5d56c184/charts/warpstream-agent/templates/deployment.yaml#L123

```yaml
deploymentKind: StatefulSet

fullnameOverride: ws

statefulSetConfig:
  clusterDomain: example.com
  
extraEnv:
- name: WARPSTREAM_TLS_ENABLED
  value: "true"
- name: WARPSTREAM_TLS_SERVER_CERT_FILE
  value: /etc/kafka/ssl/tls.crt
- name: WARPSTREAM_TLS_SERVER_PRIVATE_KEY_FILE
  value: /etc/kafka/ssl/tls.key

autoscaling:
  enabled: false

volumeMounts:
- mountPath: "/etc/kafka/ssl"
  name: warpstream-ssl
  readOnly: true

volumes:
- name: warpstream-ssl
  secret:
    secretName: warpstream-tls
```

**Comment by rmb938** [comment]
Ah, by overriding the clusterDomain, yep I see that can work. Awesome. Will approve and merge once the enable/disable flag is added :)

**Comment by avinash-platformatory** [comment]
I have added the flag. Does that look good?

**Comment by rmb938** [comment]
I only see your one commit on the PR? Not seeing any additional commits to add the variable to `values.yaml` or a if statement enabling or disabling?

**Comment by avinash-platformatory** [comment]
My bad, didn't realise the push failed. It should be there now.

**Comment by rmb938** [comment]
Perfect! Our CI only runs for maintainers of the repo so I will cherry-pick your commits into a new PR so you still get credit for contributing.

**Comment by rmb938** [comment]
merged in this PR https://github.com/warpstreamlabs/charts/pull/157 Thanks!

**Pull request closed without merging**
